Fix: KeyError error on unauthenticated API calls & persist authentication when `HTTP_REMOTE_USER` enabled

ajgon opened the pull request:
## Proposed change

### The problem

Enabling `ENABLE_HTTP_REMOTE_USER` makes all `/api/` endpoints completely useless, until the `HTTP_REMOTE_USER` header is sent. This is problematic for mobile applications, behined forward-auth guardians (like authelia), because even exposing `/api/` to them, doesn't help.

### Expected behavior

- User wants to have web app hidden behind reverse proxy with forwarding auth, and be auto-logged in via `HTTP_REMOTE_USER` header
- User also wants to use mobile application. To do so, he whitelists `/api/*` endpoints, so application can authenticate and work properly.

### Actual behavior

- Enabling `ENABLE_HTTP_REMOTE_USER`, makes all `/api/` endpoints return `500` if no remote-user header is sent, rendering all mobile apps unusable.

### Proposed solution

Unfortunatelly, I'm not a python dev, and I understand even less all magic in Django, but effectively the necessary patch (which this PR is attempting to do), should do two things:

- `/api` (root only) and `/api/token` endpoints needs to have authentication disabled completely. It doesn't make much sense anyway, as `/api/token` is used solely to authenticate, and `/api` only returns list of endpoints, which is also fine. I suggest whitelisting `/api`, because some clients (paperless-mobile for example) use it to check if server is alive. I managed to do this in dirty and hacky way in `urls.py` - it works, but I'm sure it can be done better :)
- `HttpRemoteUserMiddleware` should inherit from `PersistentRemoteUserMiddleware` instead of `RemoteUserMiddleware`. This is because, when user disabled forwarding-proxy on `/api` endpoints (to make mobile apps work), means these endpoints never receive `Remote-User` header. Since web-app uses `/api` as well, these requests will remain unauthenticated from web-app point of view. Switching to `PersistentRemoteUserMiddleware` solves this issue, as only one request with `Remote-User` header needs to be sent, for django to remember the session until logout. And conviniently - this is done, via root url of webapp, fetching main view.

Btw, probably some tests wouldn't also hurt, but as I mention - I have no idea how to do them, since I python is not my cup of tea - I'm also marking this PR as WIP, so somebody can pick it up and turn it into proper python code, as I'm unable to do it better :(

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain)

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [ ] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [ ] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [ ] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

shamoon commented:
Thanks for trying to dig into this, I am not an expert in this stuff and I also find it quite challenging to test. But:

I am not sure about your first note that api endpoints always 500 without `HTTP_REMOTE_USER`. Passing that header to paperless is done via your reverse-proxy but you didnt mention what RP youre using. For example, in my setup which uses traefik + authelia I use a different set of middlewares for the main app and for `/api/` and in fact I do not pass the auth response headers (e.g. `Remote-User` etc) **at all**. See config below. With this setup I can in fact hit `/api/token/` as expected.

Again, Im not an expert so if others have more to add I would welcome that.

Another thing wrt the PR, Im not really clear why leaving `/api/` completely open is needed (couldn't apps use /token to test server up?)

~~

```
      - traefik.http.middlewares.authelia.forwardauth.address=http://authelia:9091/api/verify?rd=https://login.domain.com/
      - traefik.http.middlewares.authelia.forwardauth.trustForwardHeader=true
      - traefik.http.middlewares.authelia.forwardauth.authResponseHeaders=Remote-User, Remote-Groups, X-Authenticated-User
```

```
      # Middlewares > chains
      - traefik.http.middlewares.chain-authelia.chain.middlewares=crowdsec-bouncer,rate-limit,corsheaders,authelia
      - traefik.http.middlewares.chain-no-auth.chain.middlewares=crowdsec-bouncer,rate-limit,corsheaders
```

```
      - traefik.http.routers.paperless.middlewares=chain-authelia
...
      - traefik.http.routers.paperless-api-rtr.rule=Host(`paperless.domain.com`) && PathPrefix(`/api/`) // other matchers removed for clarity
      - traefik.http.routers.paperless-api-rtr.middlewares=chain-no-auth
```

And:

```bash
curl -i -L -H "Authorization: Token ***" https://paperless.domain.com/api/
HTTP/2 200
date: Sat, 03 Jun 2023 05:59:35 GMT
```

ajgon commented:
I have exactly the same setup in my reverse proxy (I'm using nginx k8s ingress, but the idea is the same).

The problem is, that mobile application needs to get this token in a first place. For clarity - I'm using [paperless-mobile](https://github.com/astubenbord/paperless-mobile). In this app, authentication is done this way:

- Firstly, app pings https://paperless.mydomain.com/api/ to check if domain is alive. It does it without any token - just naked `GET` request. When `ENABLE_HTTP_REMOTE_USER` is on, and `/api/*` is "chain-no-auth", it raises `500` error:

```
$ curl https://paperless.domain.com/api/
<!doctype html>
<html lang="en">
<head>
  <title>Server Error (500)</title>
</head>
<body>
  <h1>Server Error (500)</h1><p></p>
</body>
</html>
```

and in logs I see:

```
[2023-06-03 10:15:02,989] [ERROR] [django.request] Internal Server Error: /api/
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/asgiref/sync.py", line 486, in thread_handler
    raise exc_info[1]
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/exception.py", line 43, in inner
    response = await get_response(request)
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py", line 253, in _get_response_async
    response = await wrapped_callback(
  File "/usr/local/lib/python3.9/site-packages/asgiref/sync.py", line 448, in __call__
    ret = await asyncio.wait_for(future, timeout=None)
  File "/usr/local/lib/python3.9/asyncio/tasks.py", line 442, in wait_for
    return await fut
  File "/usr/local/lib/python3.9/site-packages/asgiref/current_thread_executor.py", line 22, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.9/site-packages/asgiref/sync.py", line 490, in thread_handler
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/views/decorators/csrf.py", line 55, in wrapped_view
    return view_func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/views/generic/base.py", line 103, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 509, in dispatch
    response = self.handle_exception(exc)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 469, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
    raise exc
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 497, in dispatch
    self.initial(request, *args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 414, in initial
    self.perform_authentication(request)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 324, in perform_authentication
    request.user
  File "/usr/local/lib/python3.9/site-packages/rest_framework/request.py", line 227, in user
    self._authenticate()
  File "/usr/local/lib/python3.9/site-packages/rest_framework/request.py", line 380, in _authenticate
    user_auth_tuple = authenticator.authenticate(self)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/authentication.py", line 230, in authenticate
    user = authenticate(request=request, remote_user=request.META.get(self.header))
  File "/usr/local/lib/python3.9/site-packages/django/views/decorators/debug.py", line 42, in sensitive_variables_wrapper
    return func(*func_args, **func_kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/contrib/auth/__init__.py", line 89, in authenticate
    user_login_failed.send(
  File "/usr/local/lib/python3.9/site-packages/django/dispatch/dispatcher.py", line 176, in send
    return [
  File "/usr/local/lib/python3.9/site-packages/django/dispatch/dispatcher.py", line 177, in <listcomp>
    (receiver, receiver(signal=self, sender=sender, **named))
  File "/usr/src/paperless/src/paperless/signals.py", line 31, in handle_failed_login
    f"Login failed for user `{credentials['username']}`"
KeyError: 'username'
```

- Second step, for mobile app is to send username and password to `/api/token/` endpoint to get the token. It returns the same 500 and the same exception.

```bash
curl -X POST -d 'username=user&password=pass' https://paperless.domain.com/api/token/
```

That's why disabling auth middlewares on these two helps. 

For reference, here are my env variables:

```
PAPERLESS_ALLOWED_HOSTS: "*"
PAPERLESS_CONSUMER_POLLING: 30 # in seconds, ionotify doesn't work on ceph
PAPERLESS_DBHOST: postgresql-ha.storage.svc.cluster.local
PAPERLESS_DBNAME: paperless
PAPERLESS_DBPORT: 5432
PAPERLESS_DBSSLMODE: require
PAPERLESS_ENABLE_HTTP_REMOTE_USER: "true" # <----
PAPERLESS_HTTP_REMOTE_USER_HEADER_NAME: "HTTP_REMOTE_USER" # <----
PAPERLESS_OCR_LANGUAGE: "eng+pol"
PAPERLESS_PORT: 8000
PAPERLESS_TIKA_ENABLED: 1
PAPERLESS_TIKA_ENDPOINT: "http://tika.apps.svc.cluster.local:9998"
PAPERLESS_TIKA_GOTENBERG_ENDPOINT: "http://gotenberg.apps.svc.cluster.local:3000"
PAPERLESS_TIME_ZONE: "Europe/Warsaw"
PAPERLESS_URL: "https://paperless.domain.com"
```

I'll try to do some experimentations, and patch this `500` issue, to see if it helps, maybe that's the culrpit 🤔

ajgon commented:
Hah! The problem lies elsewhere 🤦 When I fixed `signals.py` error, everything works as expected. Looks like I approached the problem, from completely wrong side :D Let me alter the PR, to fix actual bug.

ajgon committed:
fix broken "failed login" signal

ajgon commented:
Ok, I altered my PR, and removed all this logic, as it proved to be completely unnecessary 😅 What I did instead, I fixed the `signals.py` file, to not throw exceptions when making unauthorized api calls. This would return API responses as expected.

I also kept the change of `RemoteUserMiddleware` to `PersistentRemoteUserMiddleware` for the reasons I meant in my original message.

shamoon committed:
Adds testing for unauthenticated API calls, simplify logging logic

shamoon commented:
Please take a look at my last changed, I simplified the logic (logging a user 'anonymous' is potentially misleading e.g. if someone actually creates that user) and added a test for this.

ajgon commented:
Yes, this is what I had in mind saying, I'm not a python dev and it should be adjusted :) Looks good!

shamoon commented:
Great =)

My only remaining question is with respect to the `PersistentRemoteUserMiddleware` change, it seems like that is needed e.g. when `REMOTE_USER` is only being passed on login pages, but that isn't true for -ngx.

See https://docs.djangoproject.com/en/4.2/howto/auth-remote-user/#using-remote-user-on-login-pages-only

Wonder if anyone else has any insights.

ajgon commented:
Try to test it without it (i.e. still using `RemoteUserMiddleware`) when you have `/api/` whitelisted and everything else behind authelia. When you visit paperless instance, and pass through authelia login screen, you will be asked to login again via basic auth. This is due the fact, that web-app relies on `/api/` endpoints, but because these are whitelisted, `Remote-User` header is not sent to them. Without persisted session, these will start returning 401 and browser will ask for http-basic-auh password.

shamoon reviewed (approved):
Ah, ok yea that does make sense! Ok LGTM

JigSawFr commented:
Thank you all, works better ! :)
